dart/biomechanics/OpenSimParser.cpp

```diff
@@ -1757,6 +1759,634 @@ void OpenSimParser::moveOsimMarkers(
+                    targetModel.skeleton->getPositions());
+    s_t grad = 2 * error
+               * originalModel.skeleton
+                     ->getGradientOfHeightWrtBodyScales(
+                         originalModel.skeleton->getPositions())
+                     .sum();
+    if (std::abs(grad) < 1e-4)
+      break;
+    scaleOriginalModel -= grad * 0.05;
+    // Want the original height to match the target height
+    originalModel.skeleton->setBodyScales(
+        Eigen::VectorXs::Ones(originalModel.skeleton->getNumBodyNodes() * 3)
+        * scaleOriginalModel);
+    if (verbose)
+    {
+      s_t newHeight = originalModel.skeleton->getHeight(
+          originalModel.skeleton->getPositions());
+      std::cout << "Updated original model height: " << newHeight
+                << "m at scale " << scaleOriginalModel << std::endl;
+    }
+  }
+
+  // 1. To prepare, we want to run through all the meshes on both the original
+  // and target models, and derive an equality between meshes (even if they may
+  // have different file names).
+  std::map<std::string, std::vector<std::string>> equalMeshes;
+  std::map<std::string, Eigen::Isometry3s> originalMeshWorldTransforms;
+  std::map<std::string, Eigen::Isometry3s> targetMeshWorldTransforms;
+  for (int i = 0; i < originalModel.skeleton->getNumBodyNodes(); i++)
+  {
+    dynamics::BodyNode* originalBody = originalModel.skeleton->getBodyNode(i);
+    for (auto* originalShapeNode : originalBody->getShapeNodes())
+    {
+      if (originalShapeNode->getShape()->getType()
+          == dynamics::MeshShape::getStaticType())
+      {
+        // 1.1. Get all the meshes on the original model
+        dynamics::MeshShape* originalMeshShape
+            = static_cast<dynamics::MeshShape*>(
+                originalShapeNode->getShape().get());
+        originalMeshWorldTransforms[originalMeshShape->getMeshPath()]
+            = originalShapeNode->getWorldTransform();
+        // 1.2. Now we need to compare this mesh with all the meshes on the
+        // target model
+        for (int j = 0; j < targetModel.skeleton->getNumBodyNodes(); j++)
+        {
+          dynamics::BodyNode* targetBody = targetModel.skeleton->getBodyNode(j);
+          for (auto* targetShapeNode : targetBody->getShapeNodes())
+          {
+            if (targetShapeNode->getShape()->getType()
+                == dynamics::MeshShape::getStaticType())
+            {
+              dynamics::MeshShape* targetMeshShape
+                  = static_cast<dynamics::MeshShape*>(
+                      targetShapeNode->getShape().get());
+              targetMeshWorldTransforms[targetMeshShape->getMeshPath()]
+                  = targetShapeNode->getWorldTransform();
+              // 1.3. Now we need to check if the meshes are functionally
+              // equivalent
+              bool equal = true;
+              if (originalMeshShape->getMesh()->mNumMeshes
+                  != targetMeshShape->getMesh()->mNumMeshes)
```

Comment:
If a mesh doesn't have exactly the same number of meshes/vertices, it won't establish equality? So this assumes that all models use the same set of meshes? I think that's fine since all OpenSim models basically derive from the same Geometry folder, but just checking for understanding.